### ISSUE TITLE
GCC 7 warnings

### ISSUE THREAD

**Issue by LW-archlinux** [OpenRC/openrc#148]
When building with gcc 7.1.1 on archlinux  openrc 0.27.2 triggers two new warnings, 
*-Wformat-truncation=* & *-Wimplicit-fallthrough=*

> The -Wformat-truncation=level option detects certain and likely output truncation in calls to the snprintf  family of formatted output functions. -Wformat-truncation=1 is included in -Wall and enabled without  optimization but works best with -O2 and higher.
>
> For example, the following function attempts to format an integer between 0 and 255 in hexadecimal,  including the 0x prefix, into a buffer of four characters. But since the function must always terminate output by the null character ('\0') such a buffer is only big enough to fit just one digit plus the prefix.  Therefore the snprintf call is diagnosed. To avoid the warning either use a bigger buffer or handle the function's return value which indicates whether or not its output has been truncated.
>
>    void f (unsigned x)
>    {
>      char d[4];
>      snprintf (d, sizeof d, "%#02x", x & 0xff);
>      …
>    }
>
>    warning: 'snprintf' output may be truncated before the last format character [-Wformat-truncation=]
>    note: 'snprintf' output between 3 and 5 bytes into a destination of size 4

> -Wimplicit-fallthrough warns when a switch case falls through. This warning has five different levels. The compiler is able to parse a wide range of fallthrough comments, depending on the level. It also handles control-flow statements, such as ifs. It's possible to suppress the warning by either adding a fallthrough comment, or by using a null statement: __attribute__ ((fallthrough)); (C, C++), or [[fallthrough]]; (C++17), or [[gnu::fallthrough]]; (C++11/C++14). This warning is enabled by -Wextra.

Output of  **make "${_makeargs[@]}"** is attached, make install didn't output any warnings.

Used settings : 

```bash
_makeargs=(BRANDING='Arch Linux')
_makeargs+=(MKPAM=pam)
_makeargs+=(MKSELINUX=no)
_makeargs+=(MKTERMCAP=ncurses)
_makeargs+=(PKG_PREFIX="")
_makeargs+=(LIBDIR=/usr/lib)
_makeargs+=(LIBMODE=0644) # enable binary stripping by makepkg
_makeargs+=(SHLIBDIR=/usr/lib)
_makeargs+=(LIBEXECDIR=/usr/lib/openrc)
_makeargs+=(BINDIR=/usr/bin)
_makeargs+=(SBINDIR=/usr/bin)
_makeargs+=(SYSCONFDIR=/etc/openrc) # avoid conflict with initscriptsBuild options used : for 
```

see https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=openrc for full build script .

https://gcc.gnu.org/gcc-7/changes.html

**Comment by dwfreed** [comment]
Could you provide the actual warnings you got, for those of us that don't have GCC 7 available?

**Comment by LW-archlinux** [comment]
It seems the attachment got lost somehow, so i've uploaded the log here : 
https://app.box.com/s/429ihvq7ba2ntuvu1boz4dypvgikbp5h

**Comment by dwfreed** [comment]
Github issues don't have attachments.

```
libeinfo.c: In function ‘eindent’:
libeinfo.c:897:21: warning: ‘%08d’ directive output may be truncated writing between 8 and 11 bytes into a region of size 10 [-Wformat-truncation=]
  snprintf(num, 10, "%08d", amount);
                     ^~~~
libeinfo.c:897:20: note: directive argument in the range [-2147483648, 40]
  snprintf(num, 10, "%08d", amount);
                    ^~~~~~

rc-status.c: In function ‘print_service’:
rc-status.c:115:25: warning: ‘%02ld’ directive output may be truncated writing between 2 and 18 bytes into a region of size between 0 and 28 [-Wformat-truncation=]
       "%ld day(s) %02ld:%02ld:%02ld (%s)",
                         ^~~~~
rc-status.c:115:7: note: directive argument in the range [0, 153722867280911939]
       "%ld day(s) %02ld:%02ld:%02ld (%s)",
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rc-status.c:115:7: note: directive argument in the range [-9223372036854775808, 9223372036854716339]
rc-status.c:120:20: warning: ‘%02ld’ directive output may be truncated writing between 2 and 20 bytes into a region of size between 4 and 34 [-Wformat-truncation=]
       "%02ld:%02ld:%02ld (%s)",
                    ^~~~~
rc-status.c:120:7: note: directive argument in the range [-9223372036854775808, 9223372036854716339]
       "%02ld:%02ld:%02ld (%s)",
       ^~~~~~~~~~~~~~~~~~~~~~~~
```

These are the important messages for this issue.  There are some notes about implicit fallthrough that should probably be resolved, but that's for a different issue.

**Comment by williamh** [comment]
gcc 7 is now in the stable tree on my distro, so this is fixed in the following commits:

a6165168
04721ece
c7e8f113
47e4bfae
8a945194


These will be included in 0.38.
